app/models/agents/usabilla_agent.rb

```diff
@@ -0,0 +1,236 @@
+
+    def default_options
+      {
+        'access_key' => '{% credential UsabillaAccessKey %}',
+        'secret_key' => '{% credential UsabillaSecretKey %}',
+        'expected_update_period_in_days' => '1',
+        'mode' => 'on_change',
+        'retrieve_buttons' => 'true',
+        'retrieve_apps' => 'true',
+        'retrieve_email' => 'true'
+      }
+    end
+
+    def validate_options
+      super
+
+      %w(access_key secret_key).each do |key|
+        errors.add(:base, "The '#{key}' option is required.") if options[key].blank?
+      end
+
+      if boolify(options['retrieve_buttons']).nil?
+        errors.add(:base, "The retrieve_buttons option must be true or false")
+      end
+
+      if boolify(options['retrieve_apps']).nil?
+        errors.add(:base, "The retrieve_apps option must be true or false")
+      end
+
+      if boolify(options['retrieve_email']).nil?
+        errors.add(:base, "The retrieve_email option must be true or false")
+      end
+    end
+
+    def check
+      events = []
+      events += retrieve_buttons if retrieve_buttons?
+      events += retrieve_apps if retrieve_apps?
+      events += retrieve_email if retrieve_email?
+
+      payload = events.map { |e| usabilla_response_to_event(e) }
+
+      payload.each do |e|
+        if store_payload!(previous_payloads(1), e)
+          log "Storing new result for '#{name}': #{e.inspect}"
+          create_event payload: e
+        end
+      end
+    end
+
+    private
+
+    def previous_payloads(num_events)
+      # Larger of UNIQUENESS_FACTOR * num_events and UNIQUENESS_LOOK_BACK
+      look_back = UNIQUENESS_FACTOR * num_events
+      look_back = UNIQUENESS_LOOK_BACK if look_back < UNIQUENESS_LOOK_BACK
+
+      events.order('id desc').limit(look_back) if interpolated['mode'] == 'on_change'
+    end
+
+    # This method returns true if the result should be stored as a new event.
+    # If mode is set to 'on_change', this method may return false and update an
+    # existing event to expire further in the future.
+    # Also, it will retrive asignee and/or ticket if the event should be stored.
+    def store_payload!(old_events, result)
+      case interpolated['mode'].presence
+      when 'on_change'
+        result_json = result.to_json
+        if found = old_events.find { |event| event.payload.to_json == result_json }
+          found.update!(expires_at: new_event_expiration_date)
+          false
+        else
+          true
+        end
+      when 'all', 'merge', ''
+        true
+      else
+        raise "Illegal options[mode]: #{interpolated['mode']}"
+      end
+    end
+
+    def retrieve_buttons?
+      boolify(interpolated['retrieve_buttons'])
+    end
+
+    def retrieve_apps?
+      boolify(interpolated['retrieve_apps'])
+    end
+
+    def retrieve_email?
+      boolify(interpolated['retrieve_email'])
+    end
+
+    def retrieve_buttons
+      ids = interpolated['buttons_to_retrieve'] || '*'
```

Comment:
We could add `*` as a default value in `default_options` or check here using `blank?` instead, because form submits a blank string and `usabilla_api` raises an error if form field is empty. Same for `apps_to_retrieve` and `emails_to_retrieve`.